crates/types/src/lib.rs

```diff
@@ -374,14 +374,18 @@ impl WasmFuncType {
         let non_i31_gc_ref_params_count = params
             .iter()
             .filter(|p| match **p {
-                WasmValType::Ref(rt) => rt.heap_type != WasmHeapType::I31,
+                WasmValType::Ref(rt) => {
+                    rt.heap_type != WasmHeapType::I31 && rt.heap_type != WasmHeapType::Func
+                }
                 _ => false,
             })
             .count();
         let non_i31_gc_ref_returns_count = returns
             .iter()
             .filter(|r| match **r {
-                WasmValType::Ref(rt) => rt.heap_type != WasmHeapType::I31,
+                WasmValType::Ref(rt) => {
+                    rt.heap_type != WasmHeapType::I31 && rt.heap_type != WasmHeapType::Func
+                }
                 _ => false,
             })
```

Comment:
We can reuse [this helper](https://github.com/bytecodealliance/wasmtime/blob/2a6b13295542fd00c1628880619509a6bc1c422f/crates/types/src/lib.rs#L123) here:

```suggestion
            .filter(|r| r.is_gc_heap_type())
```

(and above)